pallets/verifiers/src/tests.rs

```diff
@@ -95,32 +137,164 @@ mod register_should {
+        test_ext.execute_with(|| {
+            assert_noop!(
+                FakeVerifierPallet::register_vk(RuntimeOrigin::root(), Box::new(42)),
+                DispatchError::BadOrigin
+            );
+        })
+    }
+
+    #[rstest]
+    fn handle_double_registration_by_different_users(mut def_vk: sp_io::TestExternalities) {
+        def_vk.execute_with(|| {
+            let initial_reserved_balance = Balances::reserved_balance(USER_2);
+            assert_ok!(FakeVerifierPallet::register_vk(
+                RuntimeOrigin::signed(USER_2),
+                Box::new(REGISTERED_VK)
+            ));
+            assert!(Balances::reserved_balance(USER_2) > initial_reserved_balance);
+        })
+    }
+
+    #[rstest]
+    fn handle_double_registration_by_same_user(mut def_vk: sp_io::TestExternalities) {
+        def_vk.execute_with(|| {
+            let initial_reserved_balance = Balances::reserved_balance(USER_1);
+            assert_ok!(FakeVerifierPallet::register_vk(
+                RuntimeOrigin::signed(USER_1),
+                Box::new(REGISTERED_VK)
+            ));
+            assert_eq!(Balances::reserved_balance(USER_1), initial_reserved_balance);
+        })
+    }
+}
+
+mod unregister_should {
     use super::*;
+    use registered_vk::*;
 
-    pub const REGISTERED_VK: Vk = 4325;
-    pub const REGISTERED_VK_HASH: H256 = H256(hex!(
-        "7aeb79b96627dd87eac158bec5612ddb7f350513a179d9ab0daf4ab5788c3262"
-    ));
-    pub const VALID_HASH_REGISTERED_VK: H256 = H256(hex!(
-        "a65dc57cd8f1e436aaa8a8a473005040a4594f5411e0d9c7c5d7f20630217b79"
-    ));
+    #[rstest]
+    fn unregister_a_previously_registered_vk(mut def_vk: sp_io::TestExternalities) {
+        def_vk.execute_with(|| {
+            assert!(FakeVerifierPallet::vks(REGISTERED_VK_HASH).is_some());
+            FakeVerifierPallet::unregister_vk(RuntimeOrigin::signed(USER_1), REGISTERED_VK_HASH)
+                .unwrap();
+            assert!(FakeVerifierPallet::vks(REGISTERED_VK_HASH).is_none());
+        })
+    }
 
-    /// Provide an environment with a registered vk (use used the default vk)
-    #[fixture]
-    fn def_vk(mut test_ext: sp_io::TestExternalities) -> sp_io::TestExternalities {
-        test_ext.execute_with(|| {
-            FakeVerifierPallet::register_vk(RuntimeOrigin::signed(1), Box::new(REGISTERED_VK))
+    #[rstest]
+    fn keep_previously_registered_vk_around_if_another_user_is_referencing_it(
+        mut def_vk: sp_io::TestExternalities,
+    ) {
+        def_vk.execute_with(|| {
+            FakeVerifierPallet::register_vk(RuntimeOrigin::signed(USER_2), Box::new(REGISTERED_VK))
                 .unwrap();
-            System::reset_events();
-        });
-        test_ext
+            FakeVerifierPallet::unregister_vk(RuntimeOrigin::signed(USER_1), REGISTERED_VK_HASH)
+                .unwrap();
+            assert!(FakeVerifierPallet::vks(REGISTERED_VK_HASH).is_some());
+        })
```

Comment:
Also check emission of VkUnregistered event ?

Reply:
Addressed in https://github.com/zkVerify/zkVerify/pull/178/commits/a4eeb34edcd2f58693e2b957c85ec532be4915e0